src/Controller/Component/FlashComponent.php

```diff
@@ -83,20 +93,159 @@ public function set($message, array $options = [])
             $message = $message->getMessage();
         }
 
-        list($plugin, $element) = pluginSplit($options['element']);
+        $options['element'] = $this->_getElement($options['element']);
 
-        if ($plugin) {
-            $options['element'] = $plugin . '.Flash/' . $element;
-        } else {
-            $options['element'] = 'Flash/' . $element;
+        $sessionKey = 'Flash.' . $options['key'];
+        $messages = [];
+        $index = 0;
+
+        if ($options['reset'] === false) {
+            $messages = $this->_session->read($sessionKey);
+        }
+
+        if (!empty($messages)) {
+            end($messages);
+            $index = key($messages);
+            reset($messages);
+            $index++;
+        }
+
+        if ($index >= $this->config('limit')) {
+            array_shift($messages);
+            $this->_session->write($sessionKey, $messages);
+            $index--;
         }
 
-        $this->_session->write('Flash.' . $options['key'], [
+        $sessionKey .= '.' . $index;
+        $this->_session->write($sessionKey, [
             'message' => $message,
             'key' => $options['key'],
             'element' => $options['element'],
             'params' => $options['params']
         ]);
+
+        return $index;
+    }
+
+    /**
+     * Edit a message of the messages stack
+     *
+     * @param int $index The index of the message to edit
+     * @param string|array $message Message as a string to overwrite or an array of params
+     *   Available keys are :
+     *   - `element` The element used to render the flash message. The format is the same as
+     *   expected for the $options value of the FlashComponent::set() method
+     *   - `message` The message to overwrite
+     *   - `params` An array of variables to make available when using an element
+     * @param string $key The flash key where the message is stored
+     * @return void
+     */
+    public function edit($index, $message, $key = 'flash')
+    {
+        $sessionKey = 'Flash.' . $key . '.' . $index;
+        if (!$this->_session->check($sessionKey)) {
+            return;
```

Comment:
Is this an exceptional case? I also wonder if we shouldn't just set the new message at the given index.